core/src/main/scala/cats/free/FreeSelective.scala

```diff
@@ -0,0 +1,146 @@
+  *
+  * Free rigid selective functors. Rigidity is enforced
+  * as the definition is based on the normal form that can
+  * only be found for rigid selective functors.
+  * This is to say that not all selective functors are free.
+  * (A good thing, in most cases).
+  */
+sealed trait FreeSelective[F[_], A] {
+  type Source
+}
+object FreeSelective {
+  abstract class Pure[F[_], A] extends FreeSelective[F, A] {
+    type Source = A
+    val run: A
+  }
+  object Pure {
+    def apply[F[_], A](x: A): Pure[F, A] = new Pure[F, A] {
+      val run: A = x
+    }
+    def unapply[F[_], A](arg: Pure[F, A]): Option[A] = Option(arg.run)
+    type Aux[F[_], A, B] = Pure[F, A] {
+      type Source = B
+    }
+  }
+  abstract class Select[F[_], B] extends FreeSelective[F, B] {
+    type Source
+
+    val feab: FreeSelective[F, Either[Source, B]]
+    val fab: F[Source => B]
+  }
+  object Select {
+    type Aux[F[_], A, B] = Select[F, B] {
+      type Source = A
+    }
+  }
+
+  implicit def selective[F[_]: Functor]: Selective[FreeSelective[F, ?]] = new Selective[FreeSelective[F, ?]] {
+    override def applicative: Applicative[FreeSelective[F, ?]] = new Applicative[FreeSelective[F, ?]] {
+      override def map[A, B](fa: FreeSelective[F, A])(f: A => B): FreeSelective[F, B] =
+        fa match {
+          case x: Pure[F, A] => Pure(f(x.run))
+          case w: Select.Aux[F, A, B] =>
+            new Select[F, B] {
+              type Source = B
+              /**
+                *There is a bit of primitive recursion here that cannot be eliminated.
+                * Since all free selective expressions are finite, this is guaranteed to terminate,
+                * however it is not tail recursive, so your stack is your own.
+                **/
+              val feab: FreeSelective[F, Either[Source, B]] =
+                map(w.feab)((e: Either[w.Source, B]) => e.leftMap(f))
+
+              val fab: F[Source => B] = Functor[F].map(w.fab)(f andThen _)
+            }
+        }
+
+      override def pure[A](x: A): FreeSelective[F, A] = Pure[F, A](x)
+
+      /**
+        * The canonical apS function agreement is enforced, since these are of
+        * necessity rigid selective functors, since the normal form that
+        * backs the free defintion only holds in the rigid case.
+        */
+      override def ap[A, B](ff: FreeSelective[F, A => B])(fa: FreeSelective[F, A]): FreeSelective[F, B] =
+        select(map[A, Either[A, B]](fa)(Left.apply[A, B]))(ff)
+    }
+
+    /**
+      * *
+      * *  From the original paper:
+      * *
+      * * instance Functor f => Selective (Select f) where
+      * *   select x (Pure y) = either y id <$> x -- Generalised identity
+      * *   select x (Select y z) = Select (select (f <$> x) (g <$> y)) (h <$> z) -- Associativity
+      * *     where
+      * *       f x = Right <$> x
+      * *       g y = \a -> bimap (,a) ($a) y
+      * *       h z = uncurry z
+      */
```

Comment:
This is almost certainly not the most efficient encoding we could imagine.

Reply:
There is an alternative free construction which may be faster:

https://github.com/snowleopard/selective/blob/master/src/Control/Selective/Free.hs

It may be less stack-friendly though. Also, the representation is a function, so can't be serialised, for example.

Reply:
P.S.: I'd be curious to learn more about the trade-offs in the Scala world, so please share your thoughts about other, more Scala-friendly free selective constructions.

Reply:
If nothing else, eliminating the type member using that encoding might simplify things greatly, and increase safety. Type members are wonky. I'll see what switching to that looks like.
Most free constructions aren't serializable in Scala AFAICT anyway, so I'm not too worried about that.
Stack is tricky, might need more thinking about. I'm not entirely sure this version is very stack friendly either.

Reply:
By the way, forgot to mention that the free construction I linked to is for general selective functors (not just rigid ones), i.e. it has the advantage that one can run it with the functor `Under`, getting the underapproximation of possible effects.